compressors/rANS.py

```diff
@@ -0,0 +1,335 @@
+
+
+class rANSEncoder(DataEncoder):
+    """rANS Encoder
+
+    Detailed information in the overview
+    """
+
+    def __init__(self, freqs: Frequencies, rans_params: rANSParams):
+        """init function
+
+        Args:
+            freqs (Frequencies): frequencies for which rANS encoder needs to be designed
+            rans_params (rANSParams): global rANS hyperparameters
+        """
+        self.freqs = freqs
+        self.params = rans_params
+
+    def rans_base_encode_step(self, s, state: int):
+        """base rANS encode step
+
+        updates the state based on the input symbols s, and returns the updated state
+        """
+        f = self.freqs.frequency(s)
+        block_id = state // f
+        slot = self.freqs.cumulative_freq_dict[s] + (state % f)
+        next_state = block_id * self.freqs.total_freq + slot
+        return next_state
+
+    @cache
+    def max_state_val(self, symbol):
+        """
+        max value the state can be before calling rans_base_encode_step function
+        """
+        f = self.freqs.frequency(symbol)
+        return self.params.RANGE_FACTOR * f * (1 << self.params.NUM_BITS_OUT) - 1
+
+    def shrink_state(self, state: int, next_symbol) -> Tuple[int, BitArray]:
+        """stream out the lower bits of the state, until the state is below self.max_state_val(next_symbol)"""
+        out_bits = BitArray("")
+
+        # output bits to the stream to bring the state in the range for the next encoding
+        while state > self.max_state_val(next_symbol):
+            _bits = uint_to_bitarray(
+                state % (1 << self.params.NUM_BITS_OUT), bit_width=self.params.NUM_BITS_OUT
+            )
+            out_bits = _bits + out_bits
+            state = state >> self.params.NUM_BITS_OUT
+
+        return state, out_bits
+
+    def encode_symbol(self, s, state: int) -> Tuple[int, BitArray]:
+        """Encodes the next symbol, returns some bits and  the updated state
+
+        Args:
+            s (Any): next symbol to be encoded
+            state (int): the rANS state
+
+        Returns:
+            state (int), symbol_bitarray (BitArray):
+        """
+        # output bits to the stream so that the state is in the acceptable range
+        # [L, H] *after*the `rans_base_encode_step`
+        symbol_bitarray = BitArray("")
+        state, out_bits = self.shrink_state(state, s)
+        symbol_bitarray = out_bits + symbol_bitarray
+
+        # core encoding step
+        state = self.rans_base_encode_step(s, state)
+        return state, symbol_bitarray
+
+    def encode_block(self, data_block: DataBlock):
+        # initialize the output
+        encoded_bitarray = BitArray("")
+
+        # initialize the state
+        state = self.params.initial_state(self.freqs)
+
+        # update the state
+        for s in data_block.data_list:
+            state, symbol_bitarray = self.encode_symbol(s, state)
+            encoded_bitarray = symbol_bitarray + encoded_bitarray
```

Comment:
This prepending - is this the standard method or did we use it for its simplicity (as opposed to decoding in reverse or something else)?

Reply:
Prepending just seems very anti streaming and inefficient since it involves lots of memcopies and allocations (unless bitarray uses a deque internally)

Reply:
yes, practically encoding is performed from end to the beginning

Reply:
I was concerned about it too.. practically it seems a large memory is allocated beforehand, and then it is just filled backwards.. https://github.com/rygorous/ryg_rans/blob/c9d162d996fd600315af9ae8eb89d832576cb32d/main.cpp#L176

unfortunately there is no way around this (we have to either decode from behind or encode from behind)

Reply:
Added a comment